keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1215 @@
+| Cluster external traffic | Not supported                                                                                                    | Partially supported via IPBlock                                                    |
+| Namespace selectors      | Supports advanced selection of Namespaces with the use of `namespaceSet`                                         | Supports label based Namespace selection with the use of `namespaceSelector` field |
+
+Note that AdminNetworkPolicy can also apply to Pods in Namespaces that don't
+exist yet, and will automatically apply to a new Namespace as long as the new
+Namespace's labels match the AdminNetworkPolicy rule's appliedTo selection
+criteria. NetworkPolicies, on the contrary, only apply to Pods in the Namespace
+they are created in.
+
+### Notes/Constraints/Caveats
+
+It is important to note that the controller implementation for cluster-scoped
+policy APIs will not be provided as part of this KEP. Such controllers which
+realize the intent of these APIs will be provided by individual network policy
+providers, as is the case with the NetworkPolicy API.
+
+### Risks and Mitigation
+
+To understand why traffic between a pair of Pods is allowed or denied, a list of 
+NetworkPolicy resources in both Pods' Namespace used to be sufficient (considering 
+no other CRDs in the cluster tries to alter traffic behavior). With the introduction
+of AdminNetworkPolicy this is no longer the case, and users could face difficulty 
+in determining why NetworkPolicies did not take effect. 
+
+For example, in the case where a positive priority AdminNetworkPolicy rule,
+NetworkPolicy rule and "0" priority AdminNetworkPolicy rule apply to an overlapping
+set of Pods, users will need to refer to the priority associated with the
+rule to determine which rule would take effect. Figuring out how stacked policies
+affect traffic between workloads might not be very straightforward.
+
+To mitigate this risk and improve usability, some additional in-tree tooling
+for both the Admin and Developer will need to be created. For the Admin, it is 
+safe to assume they will have the correct RBAC roles to list all the NetworkPolicies 
+and AdminNetworkPolicies in a cluster. Therefore, the Admin oriented tooling should 
+be able to both alert the Admin to any overriding of NetworkPolicies that may occur if a 
+new AdminNetworkPolicy is to be created and provide a warning if there is another ANP 
+with the same priority. For the Developer, who usually can only list the NetworkPolicies 
+in a given namespace, the tooling should simply alert if a given NetworkPolicy would
+be overridden by any of the ANPs in a cluster. The aforementioned tooling will not 
+be a primary development goal during the alpha version of this API, and will most 
+likely be completed during the beta development cycle. 
```

Comment:
Do you foresee adding read-only access to AdminNetworPolicy to the default `view` ClusterRole`?

Not adding this makes the tooling of limited use, but once added, it's difficult / impossible to remove, thanks to backwards-compatability guarantees.

Reply:
No I don't think we would.  There will be two sets of tooling, one for the cluster-admin, and the other for the developer.  The Admin will have the ability to see exact ANP rule issues, while the Developer will only be able to be alerted that "an ANP is overriding your NP please contact your cluster-admin".

Reply:
see `risks-and-mitigations` for more

Reply:
In order for that to be the case, either that functionality needs to be built directly in to the apiserver, or developers need read access to all ANPs.

(It could also be built via an aggregated apiserver, but down that way lies madness)

Reply:
I don't think we should.  We talked about tooling a few weeks back and one of the ideas was something like a "probe" resource that a user would create and the implementation would update to indicate success or failure-because, but without leaking tenancy info.

E.g. if I am tenant=coke, I probably should not be able to learn that tenant=pepsi *exists* much less what their policies are.

Reply:
Oh, totally agreed. Just pointing out an implication of this paragraph.

(Yeah, a `/probe` resource, implemented by an aggregated apiserver, would actually be really cool. Even just for standard NP)